processor/servicegraphprocessor/processor.go

```diff
@@ -475,6 +490,47 @@ func (p *serviceGraphProcessor) storeExpirationLoop(d time.Duration) {
 	}
 }
 
+// speculate virtual node before edge get expired.
+func (p *serviceGraphProcessor) trySpeculateEvictHead(e *store.Edge) bool {
```

Comment:
This function is not really needed. It is being called in `OnExpire`, which is called by the store when an edge has expired. A lot of the code here is redundant. It can be simplified to just checking the peer labels and aggregating the edge to the metrics.